aaelf64-morello/aaelf64-morello.rst

```diff
@@ -865,3 +929,195 @@ veneer is used. The BX changes the execution state from A64 to C64:
+  copy of ``CTPIDR_EL0`` as an argument and preserve it.
+
+
+Static TLS block resolver
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+If the TLS variable is in the static block, while resolving the
+``R_MORELLO_TLSDESC`` relocation, the dynamic linker will place in the two GOT
+slots associated with this variable:
+
+- A capability to the static TLS block resolver function at offset 0.
+
+- The offset of the variable in the static TLS block at offset 16 (8 bytes).
+
+- The size of the variable at offset 24 (8 bytes).
+
+An implementation of the static block resolver could be the following:
+
+.. code-block:: text
+
+  ldp x0, x1, [c0, #16]
+  add c0, c2, x0
+  scbnds c0, c0, x1
+  ret c30
+
+Local Exec
+^^^^^^^^^^
+
+The capability to the TLS variable is derived from ``CTPIDR_EL0``. There
+are no requirements on how this is performed or the registers used, except
+that the sequence doesn't produce a dynamic relocation. A possible
+instruction sequence could be:
+
+.. code-block:: text
+
+  mrs c0, CTPIDR_EL0
+  movz x8, #:tprel_g1:local_exec_var
+  movk x8, #:tprel_g0_nc:local_exec_var
+  movz x9, #:size_g1:local_exec_var
+  movk x9, #:size_g0_nc:local_exec_var
+  add c0, c0, x8, uxtx
+  scbnds c0, c0, x9
+
+Initial Exec
+^^^^^^^^^^^^
+
+The capability to the TLS variable is derived from ``CTPIDR_EL0``. The size
+and offset of the TLS variable is stored in a GOT slot (first 8 bytes contains the
+offset and the second 8 bytes the size). This GOT slot is initialized by a
+``R_MORELLO_TPREL128`` dynamic relocation. The access must use the
+``R_MORELLO_TLSIE_ADR_GOTTPREL_PAGE20`` and ``R_MORELLO_TLSIE_ADD_LO12``
+relocations in order to allow relaxation to Local Exec. There are no other
+requirements on how this is performed or the registers used. A possible
+instruction sequence could be:
+
+.. code-block:: text
+
+  adrp c0, :gottprel:initial_exec_var
+  add c0, c0, :gottprel_lo12:initial_exec_var
+  ldp x0, x8, [c0]
+  mrs c1, CTPIDR_EL0
+  add c0, c1, x0, uxtx
+  scbnds c0, c0, x8
+
+Initial Exec to Local Exec relaxation
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The linker will generate 16 bytes in the ``.rodata`` section, containing
+the offset in the static TLS block in the first 8 bytes and the size of the
+symbol in the next 8 bytes:
+
+.. code-block:: text
+
+  .section .rodata
+  _sym_data:
+    .xword tlsoffset(sym)
+    .xword sizeof(sym)
+
+.. note::
+
+  ``tlsoffset(sym)`` denotes the offset in the static TLS block of the symbol
+  ``sym``, while sizeof(sym) denotes the size of the symbol ``sym``. These are
+  not valid assebler directives.
+
+The relaxation is performed by:
+
+- changing the ``R_MORELLO_TLSIE_ADR_GOTTPREL_PAGE20`` relocation on the symbol
+  ``sym`` to a ``R_MORELLO_ADR_PREL_PG_HI20`` with the symbol ``_sym_data``
+
+- changing the ``R_MORELLO_TLSIE_ADD_LO12`` relocation on symbol the ``sym`` to
+  a ``R_AARCH64_ADD_ABS_LO12`` relocation with the symbol ``_sym_data``.
```

Comment:
Is this requiring the addition of a symbol `_sym_data` by the linker, or is that symbol just there as an aid for explanation?
I expect the latter, but it would be nice to add a phrase to make it clearer.